File: python/y402/fastapi/request_data.py
from fastapi import Request
from starlette.routing import Match


def get_root_url(request: Request) -> str:
    """
    Retrieves the root URL (i.e. no path nor trailing slash).

    Args:
        request: The current request.

    Returns:
        The base proto://host url.
    """

    # This uses forwarded headers if present (e.g. behind reverse proxy).
    scheme = request.headers.get("x-forwarded-proto", request.url.scheme)

    # X-Forwarded-Host (and Proxy-Protocol) have precedence.
    host_header = request.headers.get("x-forwarded-host")
    if host_header:
        host = host_header
    else:
        # Fallback to the original host.
        host = request.headers.get("host") or request.url.netloc

    # Construct root URL without path/query.
    return f"{scheme}://{host}"


def resolve_endpoint(request: Request):
    """
    Obtains the endpoint (function) being responsible for
    the handing of the current request.

    Args:
        request: The request to analyze.

    Returns:
        The endpoint handling it.
    """

    for route in request.app.router.routes:
        match, _ = route.matches(request.scope)
        if match == Match.FULL:
            return getattr(route, "endpoint", None)
    return None

File: ts/y402-server-next/src/request_data.ts
import type { NextRouteContext } from "./index";

/**
 * Retrieves the root URL (i.e. no path nor trailing slash).
 * @param request The current request.
 * @returns The base proto://host URL.
 */
export function getRootUrl(request: Request): string {
    const forwardedProto = request.headers.get("x-forwarded-proto");
    const scheme = forwardedProto || new URL(request.url).protocol.replace(":", "");
    const host = request.headers.get("x-forwarded-host")
        || request.headers.get("host")
        || new URL(request.url).host;
    return `${scheme}://${host}`;
}

/**
 * Resolves one reference parameter from route context.
 * @param context The Next route context.
 * @param paramName The parameter name to read.
 * @returns A string value if present, else null.
 */
export function resolveReferenceParam(
    context: NextRouteContext,
    paramName: string
): string | null {
    const value = context.params?.[paramName];
    if (typeof value === "string") {
        return value;
    }
    if (Array.isArray(value) && value.length > 0) {
        return value[0] ?? null;
    }
    return null;
}
